src/Microsoft.OData.Core/UriParser/Resolver/AlternateKeysODataUriResolver.cs

```diff
@@ -39,20 +39,17 @@ public AlternateKeysODataUriResolver(IEdmModel model)
         /// <returns>The resolved key list.</returns>
         public override IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, Func<IEdmTypeReference, string, object> convertFunc)
         {
-            IEnumerable<KeyValuePair<string, object>> convertedPairs;
-            try
+            if (base.TryResolveKeys(type, namedValues, convertFunc, out IEnumerable<KeyValuePair<string, object>> convertedPairs))
             {
-                convertedPairs = base.ResolveKeys(type, namedValues, convertFunc);
+                return convertedPairs;
             }
-            catch (ODataException)
+
+            if (!TryResolveAlternateKeys(type, namedValues, convertFunc, out IEnumerable<KeyValuePair<string, object>> alternateConvertedPairs))
             {
-                if (!TryResolveAlternateKeys(type, namedValues, convertFunc, out convertedPairs))
-                {
-                    throw;
-                }
+                throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyOrAlternateKeyMismatch(type.FullName()));

```

Comment:
Why is `BadRequest_KeyOrAlternateKeyMismatch` not `BadRequest_AlternateKeyMismatch`?

Reply:
Because at this point, we were not able to much neither the declared key nor the alternate key.

Reply:
Is this because the previous code was incorrect, or is this because the new code is unable to differentiate? I don't see why, in the new code, the `Try` variants of the methods don't have an additional `out` parameter which is the exception that would have otherwise been throw, that way we get to keep the richness of our current errors.